File: Grokking the coding interview/Sliding Window/1. Two Sum.js
/*
Brute Force approach
find all possible substring with sum === target
TC: O(N**2), SC: O(1)
*/

var twoSum = function(nums, target) {
    
    for(let i = 0; i < nums.length; i++){
        for(let j = i + 1; j < nums.length; j++){
            if (nums[i] + nums[j] === target) return [i, j];
        }
    }
};


/*
TC: O(N), SC: O(N)
*/
var twoSum = function(nums, target) {
    let obj = {};
    
    for(let i = 0; i < nums.length; i++){
        let diff = target - nums[i];
        
        if (diff in obj) return [obj[diff], i];
        
        obj[nums[i]] = i;
    }
};

File: HashMap/1. Two Sum.py
#TC: O(N) SC: O(N)

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        obj = {}
        for ind, num in enumerate(nums):
            if (target - num) in obj:
                return [obj[target - num], ind]
            
            else:
                obj[num] = ind